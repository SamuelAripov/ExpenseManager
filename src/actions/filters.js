// SET_TEXT_FILTER
export const setTextFilter = (text = '') => ({
    type: 'SET_TEXT_FILTER',
    text
})
// SORT_BY_DATE
export const sortByDate = (date) => ({
    type: 'SORT_BY_DATE',
    date
})
// SORT_BY_AMOUNT
export const sortByAmount = (amount) => ({
    type: 'SORT_BY_AMOUNT',
    amount
})
// SET_START_DATE
export const setStartDate = (startDate = 0) => ({
    type: 'SET_START_DATE',
    startDate
})
// SET_END_DATE
export const setEndDate = (endDate = 0) => ({
    type: 'SET_END_DATE',
    endDate
})