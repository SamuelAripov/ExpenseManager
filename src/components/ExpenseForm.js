import React from 'react'
import DatePicker from 'react-datepicker'
export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? props.expense.createdAt : new Date().valueOf(),
            calendarFocused: false
        }
    }
    
    onDescripitionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onTextFormChange = (e) => {
        const textValue = e.target.value
        this.setState(() => ({ note: textValue }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }
    onSubmit = (e) => {
        e.preventDefault()
        this.props.onSubmit({
            description: this.state.description,
            amount: this.state.amount,
            createdAt: this.state.createdAt,
            note: this.state.note
        })
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescripitionChange}
                        required
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        required
                    />
                    <br />
                    <DatePicker
                        selected={this.state.createdAt}
                        onChange={this.onDateChange}
                        dateFormat="MMMM d, yyyy"
                        onChangeRaw={(e) => e.preventDefault()}
                    />
                    <textarea
                        placeholder="Add a note for you expense (optional)"
                        value={this.state.note}
                        onChange={this.onTextFormChange}
                    >
                    </textarea>
                    <button>Submit</button>
                </form>
            </div>
        )
    }
}