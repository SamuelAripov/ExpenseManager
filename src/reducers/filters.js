//Filter Reducer 
import moment from 'moment'
const filtersReducerDefaultState = {
    text: '',
    sortBy: 'date',
    startDate: moment().startOf('month').valueOf(),
    endDate: moment().endOf('month').valueOf()
}

export default (filter = filtersReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_TEXT_FILTER': {
            return {
                ...filter,
                text: action.text
            }
        }
        case 'SORT_BY_DATE': {
            return {
                ...filter,
                sortBy: 'date'
            }
        }
        case 'SORT_BY_AMOUNT': {
            return {
                ...filter,
                sortBy: 'amount'
            }
        }
        case 'SET_START_DATE': {
            return {
                ...filter,
                startDate: action.startDate
            }
        }
        case 'SET_END_DATE': {
            return {
                ...filter,
                endDate: action.endDate
            }
        }
        default:
            return filter
    }
}