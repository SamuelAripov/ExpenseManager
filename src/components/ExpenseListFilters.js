import React from 'react'
import { connect } from 'react-redux'
import DatePicker from 'react-datepicker'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
export class ExpenseListFilters extends React.Component {
    constructor(props){
        super(props)
        this.state={
            text: this.props.filters ? this.props.filters.text : '',
            sortBy: this.props.filters ? this.props.filters.sortBy : 'date',
            startDate: this.props.filters ? this.props.filters.startDate : new Date().valueOf(),
            endDate: this.props.filters ? this.props.filters.endDate : new Date().valueOf()
        }
    }
    render() {
        return (
            <div>
                <label>Search: <input type='text' value={this.props.filters.text} onChange={(e) => {this.props.setTextFilter(e.target.value)}}></input></label>
                <select onChange={(e) => {
                    e.target.value === 'date' ? this.props.sortByDate('Date') : this.props.sortByAmount('AmountRising')
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <span>Start Date</span>
                <DatePicker
                    selected={this.props.filters.startDate}
                    onChange={(e) => this.props.setStartDate(e.valueOf())}
                    dateFormat="MMMM d, yyyy"
                    onChangeRaw={(e) => e.preventDefault()}
                />
                <span>End Date</span>
                <DatePicker
                    selected={this.props.filters.endDate}
                    onChange={(e) => this.props.setEndDate(e.valueOf())}
                    dateFormat="MMMM d, yyyy"
                    onChangeRaw={(e) => e.preventDefault()}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)