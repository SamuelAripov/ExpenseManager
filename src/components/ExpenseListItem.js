import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { removeExpense } from '../actions/expenses'
import moment from "moment";

const ListItems = ({ dispatch, id, description, amount, createdAt, note }) => {
    const now = moment(createdAt)
    return (
        <div>
            <Link to={`/edit/${id}`}><h3>Expense | {description}</h3></Link>
            <p>Price | ${amount}</p>
            <p>Spent on {now.format('MMM Do YYYY')}</p>
            {note ? <p>Note: {note}</p> : <p></p>}
            <br />
        </div>
    )
}

export default ListItems
// export default connect()(ListItems)
// <button onClick={() => dispatch(removeExpense({ id }))}>Remove Expense</button>
