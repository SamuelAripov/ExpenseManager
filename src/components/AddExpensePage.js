import React from 'react'
import { addExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'

export class AddExpensePage extends React.Component {
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={(expense) => {
                        this.props.addExpense(expense)
                        this.props.history.push('/')
                    }}
                />
            </div>)
    }
}
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    }
}
export default connect(undefined, mapDispatchToProps)(AddExpensePage)