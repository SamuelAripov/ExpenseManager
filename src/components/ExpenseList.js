import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getFilteredItems from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div className='content-container'>
        {props.expenses.length !== 0 ? (
            <div className='list-header'>
                <div className='show-for-mobile'>Expenses</div>
                <div className='show-for-desktop'>Expense</div>
                <div className='show-for-desktop'>Amount</div>
            </div>
        ) : <span></span>}
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: getFilteredItems(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)