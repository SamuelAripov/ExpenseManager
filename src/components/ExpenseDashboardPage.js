import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import TotalCounter from '../components/TotalCounter'

const ExpenseDashboardPage = () => (
    <div>
        <TotalCounter/>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage