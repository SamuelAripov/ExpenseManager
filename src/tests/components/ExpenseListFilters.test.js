import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseListFilters } from '../../components/ExpenseListFilters'
import { defaultValues, providedValues } from '../fixtures/filters'

let setTextFilter, sortByAmount, setStartDate, setEndDate, wrapper

beforeEach(() => {
    setTextFilter = jest.fn()
    sortByAmount = jest.fn()
    setStartDate = jest.fn()
    setEndDate = jest.fn()
    wrapper = shallow(
        <ExpenseListFilters
            filters={defaultValues}
            setTextFilter={setTextFilter}
            sortByAmount={sortByAmount}
            setStartDate={setStartDate}
            setEndDate={setEndDate}
        />)
})

test('should render ExpenseListFilters default Values', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseListFilters with provided Values', () => {
    wrapper.setProps({
        filters: providedValues
    })
    expect(wrapper).toMatchSnapshot()
})

test('should set filters text to "rent"', () => {
    const newValue = 'rent'
    wrapper.find('input').simulate('change', { target: { newValue } })
    expect(setTextFilter).toHaveBeenCalled()
})


test('should set sortBy to "AmountRising"', () => {
    const newValue = "AmountRising"
    wrapper.find('select').simulate('change', { target: newValue })
    expect(sortByAmount).toHaveBeenLastCalledWith(newValue)
})

test('should set both start and end date to new Values', () => {
    const startDate = 2
    const endDate = 6
    wrapper.find('DatePicker').at(0).simulate('change', {
        target: startDate
    })
    wrapper.find('DatePicker').at(1).simulate('change', {
        target: endDate
    })
    expect(setStartDate).toHaveBeenLastCalledWith({ target: startDate })
    expect(setEndDate).toHaveBeenLastCalledWith({ target: endDate })
})