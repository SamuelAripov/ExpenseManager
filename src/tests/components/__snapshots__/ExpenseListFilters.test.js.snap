// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters default Values 1`] = `
<div>
  <label>
    Search: 
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <select
    onChange={[Function]}
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <span>
    Start Date
  </span>
  <DatePicker
    allowSameDay={false}
    dateFormat="MMMM d, yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onChangeRaw={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={null}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  />
  <span>
    End Date
  </span>
  <DatePicker
    allowSameDay={false}
    dateFormat="MMMM d, yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onChangeRaw={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={null}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  />
</div>
`;

exports[`should render ExpenseListFilters with provided Values 1`] = `
<div>
  <label>
    Search: 
    <input
      onChange={[Function]}
      type="text"
      value="rent"
    />
  </label>
  <select
    onChange={[Function]}
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <span>
    Start Date
  </span>
  <DatePicker
    allowSameDay={false}
    dateFormat="MMMM d, yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onChangeRaw={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={0}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  />
  <span>
    End Date
  </span>
  <DatePicker
    allowSameDay={false}
    dateFormat="MMMM d, yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onChangeRaw={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={3}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  />
</div>
`;
